File: src/main/java/frc/robot/subsystems/GrabberSubsystem.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import com.revrobotics.CANSparkMax;
import com.revrobotics.SparkMaxPIDController;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.wpilibj.DoubleSolenoid;
import static edu.wpi.first.wpilibj.DoubleSolenoid.Value.*;
import edu.wpi.first.wpilibj.PneumaticsModuleType;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants.Grabber;

public class  GrabberSubsystem extends SubsystemBase
{

  private final CANSparkMax leftMotor;
  private final CANSparkMax rightMotor;
  private boolean in = true;
  private final DoubleSolenoid solenoid;

  public GrabberSubsystem(int rightMotorId, 
                int leftMotorId,
                int extendChannel, 
                int retractChannel
                ) {
    leftMotor = new CANSparkMax(leftMotorId,MotorType.kBrushless);
    leftMotor.setSmartCurrentLimit(10);
    rightMotor = new CANSparkMax(rightMotorId, MotorType.kBrushless);
    rightMotor.setSmartCurrentLimit(10);

    solenoid = new DoubleSolenoid(PneumaticsModuleType.REVPH, extendChannel, retractChannel);
  }

  // currently using percentage voltage might be better to change it to explicit speed
  public void setMotorsSpeeds(double speed, boolean in) {
    leftMotor.set(-speed);
    rightMotor.set(speed);
    this.in = in;
  }

  public void periodic() { 
    double current = leftMotor.getOutputCurrent();
    if (current >= Grabber.offCurrent && Grabber.disableMotors && in) {
      stopMotors();
    }
  }

  public void stopMotors() {
    leftMotor.stopMotor();
    rightMotor.stopMotor();
  }

  public void toggleMotorDirection() {
    setMotorsSpeeds( !in ? Grabber.motorSpeeds : -Grabber.motorSpeeds, !in); 
  }

  public void clamp() {
    solenoid.set(kReverse);
  }

  public void unclamp() {
    solenoid.set(kForward);
  }
}

File: src/main/java/frc/robot/subsystems/GrabberSubsystem.kt
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.
package frc.robot.subsystems

import edu.wpi.first.wpilibj.DoubleSolenoid
import edu.wpi.first.wpilibj.PneumaticsModuleType
import edu.wpi.first.wpilibj2.command.SubsystemBase
import frc.robot.Constants.Grabber

class GrabberSubsystem(rightMotorId: Int,
                       leftMotorId: Int,
                       extendChannel: Int,
                       retractChannel: Int
) : SubsystemBase() {
    // private final CANSparkMax leftMotor;
    // private final CANSparkMax rightMotor;
    private var `in` = true
    private val solenoid: DoubleSolenoid

    init {
        // leftMotor = new CANSparkMax(leftMotorId,MotorType.kBrushless);
        // leftMotor.setSmartCurrentLimit(10);
        // rightMotor = new CANSparkMax(rightMotorId, MotorType.kBrushless);
        // rightMotor.setSmartCurrentLimit(10);
        solenoid = DoubleSolenoid(PneumaticsModuleType.REVPH, extendChannel, retractChannel)
    }

    // currently using percentage voltage might be better to change it to explicit speed
    fun setMotorsSpeeds(speed: Double, `in`: Boolean) {
        // leftMotor.set(-speed);
        // rightMotor.set(speed);
        this.`in` = `in`
    }

    override fun periodic() {
        // double current = leftMotor.getOutputCurrent();
        // if (current >= Grabber.offCurrent && Grabber.disableMotors && in) {
        // stopMotors();
        // }
    }

    fun stopMotors() {
        // leftMotor.stopMotor();
        // rightMotor.stopMotor();
    }

    fun toggleMotorDirection() {
        setMotorsSpeeds(if (!`in`) Grabber.motorSpeeds else -Grabber.motorSpeeds, !`in`)
    }

    fun clamp() {
        solenoid.set(DoubleSolenoid.Value.kReverse)
    }

    fun unclamp() {
        solenoid.set(DoubleSolenoid.Value.kForward)
    }

    fun zeroMotors() {
        setMotorsSpeeds(0.0, true)
    }
}
